test/code/plugins/oms_diag_test.rb

```diff
@@ -0,0 +1,243 @@
+            assert_equal(Fluent::DiagTest::LOG_STR_04,
+                         x[OMS::Diag::DI_KEY_LOGMESSAGE],
+                         "The emitted log did not match with one intended")
+            assert_equal(OMS::Diag::DEFAULT_IPNAME,
+                         x[OMS::Diag::DI_KEY_IPNAME],
+                         "The ipname does not match with the one intended")
+        end
+    end
+
+    # Test cases pertaining to dataitem processing
+
+    # Checking if ProcessDataItemsPostAggregation is working correctly
+    def test_case_dataitem_proc_01
+        agent_id = 'test-agent-id-1234'
+        # Creating valid dataitems
+        dataitems = Array.new
+        (1..5).each do
+            dataitem = Hash.new
+            dataitem[OMS::Diag::DI_KEY_LOGMESSAGE] = 'Just a test string'
+            dataitem[OMS::Diag::DI_KEY_IPNAME] = OMS::Diag::DEFAULT_IPNAME
+            dataitem[OMS::Diag::DI_KEY_TIME] = Time.now.utc.strftime("%Y-%m-%dT%H:%M:%S.%6NZ")
+            # Randomly add a few optional properties
+            if rand(100) % 2 == 0
+                (1..3).each do
+                    dataitem[SecureRandom.uuid] = SecureRandom.uuid
+                end
+            end
+            dataitems << dataitem
+        end
+        # Calling to process the dataitems
+        OMS::Diag.ProcessDataItemsPostAggregation(dataitems, agent_id)
+
+        # Verifying the dataitems were processed right
+        for x in dataitems
+            # Check if IPName is absent
+            assert_equal(false,
+                         x.key?(OMS::Diag::DI_KEY_IPNAME),
+                         "#{OMS::Diag::DI_KEY_IPNAME} should be deleted from dataitem")
+            # Check if Agent guid is present
+            assert_equal(true,
+                         x.key?(OMS::Diag::DI_KEY_AGENTGUID),
+                         "#{OMS::Diag::DI_KEY_AGENTGUID} should be added to dataitem")
+            # Check if data item type is present
+            assert_equal(true,
+                         x.key?(OMS::Diag::DI_KEY_TYPE),
+                         "#{OMS::Diag::DI_KEY_TYPE} should be added to dataitem")
+        end
+    end
+
+    # Checking if spurious dataitems are removed by ProcessDataItemsPostAggregation
+    def test_case_dataitem_proc_02
+        agent_id = 'test-agent-id-1234'
+        # Creating valid dataitems
+        dataitems = Array.new
+        for y in 1..5 do
+            dataitem = Hash.new
+            dataitem[OMS::Diag::DI_KEY_LOGMESSAGE] = 'Just a test string'
+            dataitem[OMS::Diag::DI_KEY_IPNAME] = OMS::Diag::DEFAULT_IPNAME
+            dataitem[OMS::Diag::DI_KEY_TIME] = Time.now.utc.strftime("%Y-%m-%dT%H:%M:%S.%6NZ")
+            # Create few spurious cases
+            if y == 1
+                dataitem.delete(OMS::Diag::DI_KEY_LOGMESSAGE)
+            elsif y == 3
+                # note this is valid spurious here but without ipname
+                # write call of out_oms_diag will set it to default
```

Comment:
To elaborate on this comment.

- A dataitem of diagnostic log first goes from emission into the write call of output plugin as part of buffer chunk.
- There if there is no ipname then we assign that dataitem the default ipname (Diagnostics).
- The check is made later after aggregation by ipname in ProcessDataItemsPostAggregation for absence of ipname, where it will discard the dataitem (if missing ipname key)

- However while testing out_oms_diag write function, a spurious dataitem to it cannot be one that is missing ipname as it overrides with default ipname if it is missing